### ISSUE TITLE
Optimization of database creation

### ISSUE THREAD

**Issue by yarden** [daler/gffutils#20]
Is it possible to optimize database creation? It'd be interesting to see where the bottlenecks are. Would Cythonizing certain parts of the code help with this, or is the bottleneck is purely at the sqlite interface? For example, creation of a database for this GFF (Ensembl mouse genes) takes ~15 minutes:

```
http://genes.mit.edu/yarden/Mus_musculus.NCBIM37.65.gff
$ gffutils-cli create Mus_musculus.NCBIM37.65.gff
```

The bottleneck is there even when dbs are created in memory, though it's of course smaller in these cases. This is related to another issue we discussed which is: how to create derivative GFF files from an existing ones, i.e. how to iterate through a GFF and make a new version of it. For example, if you wanted to iterate through the above GFF and simply add an optional key, value pair to the attributes of some records. This is a kind of "streaming" operation that can be done line-by-line and doesn't necessarily need a db. The overhead of creating the db makes gffutils impractical for these kinds of simple operations on large-ish GFF files.

There are more sophisticated operations (i.e. non-streaming ones) where a naive in memory solution still is considerably faster because of the database creation bottleneck. For example, the naive GFF parser used in `misopy.gff_utils` (see `load_genes_from_gff` in https://github.com/yarden/MISO/blob/fastmiso/misopy/Gene.py) simply iterates through the GFF file multiple times to collect all the gene entries into a simple `Gene` object, with mRNAs represented as lists of exons. This kind of gene-centric in-memory solution is less expressive than gffutils (does not handle arbitrary nesting, and ignores non-genes basically) but for simple operations like "Add attribute X to all genes" or "Reformat the IDs of this GFF" it's considerably faster. It's not actually faster once the DB is created; gffutils retrieval is excellent, but the overhead of creation of the db trumps the computation time for many of these operations.

In summary, I'm wondering if there's a way to try to bridge the gap between the fast, but hacky solutions and the sophisticated gffutils solution that comes with an overhead. I think this is an important issue because many of the operations done on GFFs (at least that I do) don't require many hierarchical SQL queries.

**Comment by pavo** [comment]
Any luck with this? I am trying to graduate from the examples to the human genome (eg ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.gtf.gz) and some ChipSeq data downloaded from the SRA.  Do you have any advice on a dataset or database to try?  I don't mind waiting several days for this but hoped I could get started on something in the mean time.  Especially if it might crash and I would have to start over again :).   Also, I am naively wondering if the database creation can be parallelized - we have a 32 core workstation and database creation function only uses one CPU at 99%, memory at 0.1%, and 30 other CPUs at 0%.  Very new to databases, so maybe it is obviously not feasible but just curious!  Thanks for your help.

**Comment by daler** [comment]
I was never able to reproduce the error. It could have been a hardware-related issue, since sqlite3 can be slow on NFS filesystems. I will run the 84 Ensembl GTF file and report back here.

**Comment by daler** [comment]
Also, you mention ChIP-seq data -- is that related to this db creation issue in some way?

**Comment by daler** [comment]
Are you using the `disable_infer_genes` and `disable_infer_transcripts` kwargs? The following code:

```python
# using:
# http://ftp.ensembl.org/pub/current_gtf/homo_sapiens/Homo_sapiens.GRCh38.84.chr.gtf.gz
import gffutils
gffutils.create_db(
    'Homo_sapiens.GRCh38.84.chr.gtf.gz',
    'Homo_sapiens.GRCh38.84.chr.gtf.db',
    disable_infer_genes=True,
    disable_infer_transcripts=True)
```

took under half an hour:

```
real	24m56.864s
user	18m54.308s
sys     0m39.158s
```

Can you provide the exact commands you're using?

**Comment by pavo** [comment]
I did this:

```
import gffutils
fn = "Homo_sapiens.GRCh38.84.chr.gtf"  # this was gunzipped into the working directory
db = gffutils.create_db(fn, dbfn='test.db', force=True, keep_order=True, merge_strategy='merge', sort_attribute_values=True)

```
It looks like my processes finished overnight.  I did not set `disable_infer_genes=True and disable_infer_transcripts=True` as you did in your code.  I naively thought that I would want the gene information and transcript information, so I had kept these on.   Is it ok to use the databases I created overnight?  

Any idea why with my code the database size ramped up quickly to a 2.3GB db then basically remained the same size while the CPU worked all night?  It sounds like inferring genes and transcripts is what took such a long time so was wondering what that computation was (is it a sort operation?).  The final database size this morning is 2.4GB, not much different from the 2.3GB it ramped up to in about 20-30 minutes.  I thought that the create_db() function was simply packaging the GTF file's column data into a sql database but clearly there is more computation going on!

**Comment by daler** [comment]
You want `disable_infer_transcripts=True` and `disable_infer_genes=True` for this file. Ensembl's GTFs are not true GTFs since they contain lines for genes and transcripts already. If you're using a recent version of `gffutils`, you should get a warning saying that there were gene/transcripts in your GTF already and that it could take a while if you don't disable inferring genes/transcripts.

Putting the GTF into the features table and relations table is just the first step. When `gffutils` detects a GTF file, it does a lot of work to infer transcripts and genes.  After it infers each gene or transcript, it sees that there is already a transcript or gene in the database with the same ID. That's because this file contains genes and transcripts, so they've already been imported by this point. Then gffutils does more work to merge the information from the gene or transcript it just inferred with the one that already exists in the database.

So the extra time is due to merge operations caused by name collisions, which in turn were caused from inferring genes and transcripts that already exist in the GTF file. I'd re-run using those `disable_infer_*` args. It shouldn't take long, and if it takes longer than an hour let me know.

**Comment by iriet** [comment]
Hi @daler,

I am trying to use gffutils to work with the GRCh38 RefSeq annotations, and the database creation is taking quite some time. I am hoping you could throw us a few tips on options we should be flipping on/off. The goal of our use case is to parse exon coordinates for alternative splice variants, referenced to their genomic locations. We would like to keep the affiliations of alternate variants with their primaries, and coordinate annotations for 5'/3' UTRs too, ideally.  

The gff file was obtained from here:

ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_genomic.gff.gz

Our attempts ran for a couple of hours on my student's machine before we aborted the process. Then, I also tried it for an hour before aborting, on my (newer) 2.9 GHz Intel Core i7 laptop. The latter is running macOS Sierra 10.12.5 with gffutils version 0.8.7.1 which was installed using conda version 4.3.18. The python is 3.6.0 and the sqlite is version 3.13.0.

It would perhaps be helpful for many of us who have encountered file/source specific performance issues if there was some feedback provided about how the database creation is going, and how long it might be expected to take to reach completion. Our attempts gave no clear indication of satisfactory progress, though if we knew it would be done in a couple of hours, we would have let it keep going. On the other hand, if it is going to take days, I would happily try some of the options you have suggested above (or any others you might have, for dealing with this particular file).

thanks for any help,
Takeshi

**Comment by daler** [comment]
It should take <15 mins. Are you using `disable_infer_genes=True` and `disable_infer_transcripts=True` as described above?

**Comment by daler** [comment]
Wait, sorry, I thought this was related to the above comments. The file you link to is a GFF, so there is no gene or transcript inference by default. It appears the issue with this file is the duplicated IDs.

Databases need a unique ID for each feature. The `ID` field in this file is not unique -- for example here are the features for `ID=cds1`:

```
NC_000001.11	Gnomon	CDS	182709	182746	.	+	0	ID=cds1;Parent=rna37;Dbxref=GeneID:102725121,Genbank:XP_011542110.1;Name=XP_011542110.1;gbkey=CDS;gene=LOC102725121;product=uncharacterized protein LOC102725121 isoform X1;protein_id=XP_011542110.1
NC_000001.11	Gnomon	CDS	183114	183240	.	+	1	ID=cds1;Parent=rna37;Dbxref=GeneID:102725121,Genbank:XP_011542110.1;Name=XP_011542110.1;gbkey=CDS;gene=LOC102725121;product=uncharacterized protein LOC102725121 isoform X1;protein_id=XP_011542110.1
NC_000001.11	Gnomon	CDS	183922	184158	.	+	0	ID=cds1;Parent=rna37;Dbxref=GeneID:102725121,Genbank:XP_011542110.1;Name=XP_011542110.1;gbkey=CDS;gene=LOC102725121;product=uncharacterized protein LOC102725121 isoform X1;protein_id=XP_011542110.1
```

If you use `merge_strategy="merge"`, then gffutils assumes the lines refer to the same feature and so does a lot of work to merge the attributes in a nice way. Looking at this file though, the CDSs should definitely be considered different features.

You'll need to decide how you want to be able to refer to CDSs. If you don't really have an opinion on that, you can try the `merge_strategy="create_unique"` argument when creating the database which should speed things up considerably. The features above will then be called `cds1`, `cds1.1`, and `cds1.2`. Alternatively you can write a transform function to do arbitrary manipulation of the features before they get into the db, for example to create your own custom ID field based on the other attributes.

Give `merge_strategy="create_unique"` a try to see if it helps the speed issue. It still should run in <15 mins.

**Comment by iriet** [comment]
The merge_strategy="create_unique" worked well for what we want. Thank you kindly for the pointer.

Best
Takeshi
